Habitica/src/main/java/com/habitrpg/android/habitica/ui/adapter/social/challenges/ChallengeTasksRecyclerViewAdapter.java

```diff
@@ -0,0 +1,194 @@
+        component.inject(this);
+    }
+
+    @Override
+    public boolean loadFromDatabase() {
+        return false;
+    }
+
+    @Override
+    public int getItemViewType(int position) {
+        Task task = this.filteredContent.get(position);
+
+        if (task.type.equals(Task.TYPE_HABIT))
+            return TYPE_HABIT;
+
+        if (task.type.equals(Task.TYPE_DAILY))
+            return TYPE_DAILY;
+
+        if (task.type.equals(Task.TYPE_TODO))
+            return TYPE_TODO;
+
+        if (task.type.equals(Task.TYPE_REWARD))
+            return TYPE_REWARD;
+
+        if (addItemCallback != null && task.type.equals(TASK_TYPE_ADD_ITEM))
+            return TYPE_ADD_ITEM;
+
+        return TYPE_HEADER;
+    }
+
+    public void enableAddItem(Action1<Task> cb) {
+        addItemCallback = cb;
+    }
+
+    public int addTaskUnder(Task taskToAdd, Task taskAbove) {
+        int position = $.findIndex(this.content, t -> t.getId().equals(taskAbove.getId()));
+
+        content.add(position + 1, taskToAdd);
+        filter();
+
+        return position;
+    }
+
+    @Override
+    public BaseTaskViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
+        BaseTaskViewHolder viewHolder = null;
+
+        switch (viewType) {
+            case TYPE_HABIT:
+                viewHolder = new HabitViewHolder(getContentView(parent, R.layout.habit_item_card));
+                break;
+            case TYPE_DAILY:
+                viewHolder = new DailyViewHolder(getContentView(parent, R.layout.daily_item_card), dailyResetOffset);
+                break;
+            case TYPE_TODO:
+                viewHolder = new TodoViewHolder(getContentView(parent, R.layout.todo_item_card));
+                break;
+            case TYPE_REWARD:
+                viewHolder = new RewardViewHolder(getContentView(parent, R.layout.reward_item_card));
+                break;
+            case TYPE_ADD_ITEM:
+                viewHolder = new AddItemViewHolder(getContentView(parent, R.layout.challenge_add_task_item), addItemCallback);
+                break;
+            default:
+                viewHolder = new DividerViewHolder(getContentView(parent, R.layout.challenge_task_divider));
+                break;
+        }
+
+        viewHolder.setDisabled(openTaskDisabled, taskActionsDisabled);
+        return viewHolder;
+    }
+
+    public List<Task> getTaskList(){
+        return $.map(content, t -> t);
+    }
+
+
+    /**
+     * @param task
+     * @return true if task found&updated
+     */
+    public boolean replaceTask(Task task) {
+        int i;
+        for (i = 0; i < this.content.size(); ++i) {
+            if (content.get(i).getId().equals(task.getId())) {
+                break;
+            }
+        }
+        if (i < content.size()) {
+            content.set(i, task);
+
+            filter();
+            return true;
+        }
+
+        return false;
+    }
+
+    public class AddItemViewHolder extends BaseTaskViewHolder {
```

Comment:
so that it can be used in `SortableTasksRecyclerViewAdapter<BaseTaskViewHolder>`  or rather `BaseTasksRecyclerViewAdapter` 

(I'll remove `SortableTasksRecyclerViewAdapter` again, or do we plan to move the Challenge Items inside one list too?)